Adds Stuns & slipping

Aranclanos opened the pull request:
### Purpose
Adds Stuns
Adds Slipping
Adds wet floors
Fixes #1644 


### Open Questions and Pre-Merge TODOs

- [X]  This fix is tested on the same branch it is PR'ed to.
- [X]  I correctly commented my code
- [X]  My code is indented with tabs and not spaces
- [X]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [X]  This PR does not bring up any new compile errors
- [X]  This PR has been tested in editor
- [X]  This PR has been tested in multiplayer (with 2 clients, if applicable)

### Notes:
Renamed Prefabs/Blood to Prefabs/Decal, and added a WaterSplat prefab into Prefabs/Decal/resources.
Added the CrossedBehavior folder to Scrips/Player, and added a Slippery.cs file inside it. This Slippery.cs file can be added to any item gameobject to make it slippery.
Updated Mop code and prefab.
Updated Bloodsplat code, and added WaterSplat code. They both now inherit from FloorDecal.cs.
Updated PlayerSync.Server.cs to support item crossing behaviour.
Updated RegisterPlayer.cs to support stuns.
Updated RegisterItem.cs to support crossing behaviour.
Updated MopTrigger to create water splats (wet floors) on non-space tiles.
Added slippery.cs as a component to slippery items (banana peels and wet floor decals).
Updated EffectsFactory.cs to support water decal tiles.
Updated SpriteManager.cs to support water decal tiles.
Updated LivingHealthBehaviour.cs to support stuns.
Added Slip to the sound manager.
Added stun timers in LivingHealthBehaviour.cs.

Itsnotjames committed:
Added Slip and Stun code, mops and wet floor decals, and item crossing behaviour.

Renamed Prefabs/Blood to Prefabs/Decal, and added a WaterSplat prefab into Prefabs/Decal/resources.
Added the CrossedBehaviour folder to Scripts/Player, and added a Slippery.cs file inside it. This Slippery.cs script can be added to any item gameobject to make it slippery.
Updated Mop code and prefab.
Updated Bloodsplat code, and added WaterSplat code. They both now inherit from FloorDecal.cs.
Updated PlayerSync.Server.cs to support item crossing behaviour.
Updated RegisterPlayer.cs to support stuns.
Updated RegisterItem.cs to support crossing behaviour.
Updated MopTrigger to create water splats (wet floors) on non-space tiles.
Added slippery.cs as a component to slippery items (banana peels and wet floor decals).
Updated EffectsFactory.cs to support water decal tiles.
Updated SpriteManager.cs to support water decal tiles.
Updated LivingHealthBehaviour.cs to support stuns.

Itsnotjames committed:
various fixes

slip sound, spriterend on bloodsplat, remove using UnityEditorInternal which prevented builds from being made.

fomalsd commented on UnityProject/Assets/Scripts/Items/Tools/MopTrigger.cs:
```diff
@@ -12,7 +13,7 @@ public override bool Interact (GameObject originator, Vector3 position, string h
             return base.Interact (originator, position, hand);
         }
-        var targetWorldPos = Camera.main.ScreenToWorldPoint(CommonInput.mousePosition);
+        var targetWorldPos = Camera.main.ScreenToWorldPoint(Input.mousePosition);
```
No, use CommonInput

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterPlayer.cs:
```diff
@@ -37,36 +41,11 @@ public override bool IsPassable(Vector3Int from)
-		}
-	}
-
-	void RemainUpright()
```
Looks like you're rolling back quite a few features. Did you discard all new code when resolving merge conflicts?

Itsnotjames replied:
my bad

Itsnotjames committed:
Fixed merge conflict mistake

Re-added the code in LivingHealthBehaviour.cs, MopTrigger.cs, and RegisterPlayer.cs

Aranclanos committed:
Merge pull request #1 from Itsnotjames/develop

Fixed merge conflict mistake

fomalsd commented on UnityProject/Assets/Scripts/Player/PlayerSync.Server.cs:
```diff
@@ -596,4 +597,29 @@ private void CheckMovementServer()
+		List<GameObject> objects = UITileList.GetItemsAtPosition(position);
+		// Removes player from object list
+		objects.Remove(gameObject);
+		RegisterPlayer registerPlayer = GetComponent<RegisterPlayer>();
```
Calling GetComponent<> each time you cross something is inefficient. Cache this component into a field in init method like Start() and refer to it

Itsnotjames replied:
Did this in the next commit.

fomalsd commented on UnityProject/Assets/Scenes/Lobby.unity:
```diff
@@ -5285,7 +5285,7 @@ PrefabInstance:
         type: 3}
       propertyPath: m_AnchoredPosition.y
-      value: 409.29996
+      value: 409.30002
```
Exclude scenes, prefabs and metafiles from commits if you didn't explicitly modify them

Itsnotjames replied:
My bad

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterPlayer.cs:
```diff
@@ -118,4 +120,32 @@ public void GetUp()
+	{
+		isStunned = true;
+		LayDown();
+		playerScript.playerNetworkActions.DropItem("leftHand");
```
it'd be nice if you made object dropping optional (add a bool parameter defaulting to true)

Itsnotjames replied:
Added this.

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterItem.cs:
```diff
@@ -1,7 +1,22 @@
+
+		public void Cross(ref RegisterPlayer registerPlayer)
+		{
+			CrossedRegisterPlayer = registerPlayer;
```
IMO storing it in a field is unnecessary, also do use UnityEvents instead of delegates and c# events, it's what we use throughout the codebase.
Just define `public class RegisterPlayerEvent : UnityEvent<RegisterPlayer>{}` somewhere and use it as event field type instead. Invocation would be like `OnCrossed?.Invoke(registerPlayer)` and that's it. 
Also not sure why you use `ref` here as well

Itsnotjames replied:
Changed it to a unityevent and removed ref.

fomalsd replied:
I mean, you don't need a `RegisterPlayer CrossedRegisterPlayer` field at all in this class. just use parametrized custom UnityEvent<RegisterPlayer> that would pass player to the invoked method, Slip() in your case

fomalsd commented on UnityProject/Assets/Scripts/Health/Living/LivingHealthBehaviour.cs:
```diff
@@ -400,6 +405,26 @@ protected void CalculateOverallHealth()
 
+	[Server]
+	protected void CalculateStun(){
+		if(StunDuration > 0)
```
is there a particular reason why you chose Update() over a coroutine? Coroutines are widely used in our codebase and are the right thing to use when you want cooldowns

Itsnotjames replied:
Wasn't sure what to do here, saw how some systems used the UpdateManager's events for timers so I decided to use that here.

fomalsd commented on UnityProject/Assets/Scripts/Health/Living/LivingHealthBehaviour.cs:
```diff
@@ -143,6 +146,7 @@ public override void OnStartServer()
 		DNABloodTypeJSON = JsonUtility.ToJson(DNABloodType);
 		bloodSystem.SetBloodType(DNABloodType);
 		base.OnStartServer();
+		registerPlayer = GetComponent<RegisterPlayer>();
```
you know, not all LivingHealthBehaviours are players. you'll get NREs with animals if you try to calculate stun below

Itsnotjames replied:
Moved all that logic to RegisterPlayer, because of it's dependency on logic there like LayDown(), GetUp() and UpdateMe().

fomalsd reviewed (changes requested):
Thanks for contributing!
That's a nice feature, complex one though because of a lot of nuances.

Banana peels mostly work as intended, an exception is that they make you slip even if they're in the air and sometimes make thrower slip if he's running and throwing in the same direction. I don't think that should be the case. Check if peel is being thrown (in the air) at the moment (`customNetTransform.IsBeingThrown` on serverside) before invoking slip action.
Oh, and you probably shouldn't slip on peels in space:
![GIF](https://user-images.githubusercontent.com/10403536/54905445-f281b500-4ed8-11e9-9673-072144cbc902.gif)

Slippery tiles though should have no visual effects and no client should know in advance if floor is wet or lubed.
Also gameObject-based decals are being phased out soon for multiple reasons (#1583) so this might've needed a refactor soon. 
That's why I was mentioning MetaDataNodes in discord: mopping floor shouldn't spawn any new objects and just update floor tile's metadata to mark it as slippery.
Interacting with tile metadata might be trickier than with gameObject though.

There is another wish for slipping inertia (to make players slide through the hallway and out into the airlock), but I suppose this could be done later, it's just me being perfectionist
Also please check out the rest of comments

Itsnotjames committed:
Various fixes and changes to slips and stuns

Made object dropping optional.
Moved the registerPlayer cache to the start in PlayerSync.server.cs.
Removed ref in RegisterItem.cs.
Changed event in RegisterItem.cs to unityevent.
No longer slip on banana peels in space.
Moved stuns out of LivingHealthBehaviour, stuns are now entirely in RegisterPlayer.cs.
Disabled wet floor decal and replaced it with a IsSlippery boolean in the tile's MetaData.

Aranclanos committed:
Merge pull request #2 from Itsnotjames/hamishthree

Various fixes and changes to slips and stuns

Aranclanos committed:
Restores lobby.unity changes to develop branch

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterPlayer.cs:
```diff
@@ -118,4 +136,79 @@ public void GetUp()
+
+	public void CheckTileSlip()
+	{
+		if (metaDataLayer.IsSlipperyAt(transform.position.CutToInt()))
```
```suggestion
		if (metaDataLayer.IsSlipperyAt(Position))
```

Itsnotjames replied:
![HhlrpIk](https://user-images.githubusercontent.com/28093344/55063587-26c5c300-50dd-11e9-9d82-facd3f0b62dc.png)
Position is off by 1 on the X and Y axis, so this change makes the top right tile of the tile cleaned slippery instead.
Since Position is read in 21 different places, this is seems pretty strange to me.

fomalsd commented on UnityProject/Assets/Scripts/Player/CrossedBehaviors/Slippery.cs:
```diff
@@ -0,0 +1,33 @@
+
+	private void Slip()
+	{
+		if (MatrixManager.IsSpaceAt(transform.position.CutToInt()))
```
don't use transform.position for any serverside calculations (and preferably clientside, too)
you've got RegisterTile reference, use it instead
```suggestion
		if (MatrixManager.IsSpaceAt(registerItem.WorldPosition))
```

fomalsd commented on UnityProject/Assets/Scripts/Player/PlayerSync.Server.cs:
```diff
@@ -596,4 +600,30 @@ private void CheckMovementServer()
+		{
+			return;
+		}
+		List<GameObject> objects = UITileList.GetItemsAtPosition(position);
```
don't use UITileList, it's crap
just use MatrixManager.GetAt<RegisterItem>(pos) and you'll get a nice list of RegisterItems, no need to call GetComponent<> yourself

Jibbajabbafic replied:
Could we deprecate UITileList so people stop using it? Perhaps we could add a warning to it

fomalsd replied:
yeah, totally

fomalsd reviewed (changes requested):
please resolve things I mentioned, the rest of code looks good!
yet to test it in game though

fomalsd commented on UnityProject/Assets/Scripts/Items/Tools/MopTrigger.cs:
```diff
@@ -1,10 +1,18 @@
+
+	private void Awake()
+	{
+		metaDataLayer = transform.GetComponentInParent<MetaDataLayer>();
```
oops, you shouldn't cache metaDataLayer, it changes when item changes matrix

Itsnotjames committed:
Various changes and broken LocalPosition change.

In RegisterItem.cs changed OnCrossed unity event to use a RegisterPlayer parameter.
In RegisterPlayer.cs used correct Position Vector3Int for CheckTileSlip().
In Slippery.cs used registerItem's world position instead of calling transform.position.
In PlayerSync.Server.cs in Cross(position) used MatrixManager to get the items at a position instead of UITileList.

Itsnotjames committed:
Fixed LocalPosInt and moved CheckTileSlip()

Aranclanos committed:
Merge pull request #3 from Itsnotjames/hamishthree

Various changes

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs:
```diff
@@ -57,6 +58,31 @@ public bool ExistsAt(Vector3Int position)
+	{
+		yield return new WaitForSeconds(15f);
+		tile.IsSlippery = false;
+		Debug.Log($"Hamish: wet floor dried up at X:{tile.Position.x} Y:{tile.Position.y}");
```
either remove it or change to Logger.Log ans assign logger category

fomalsd commented on UnityProject/Assets/Scripts/Items/Tools/MopTrigger.cs:
```diff
@@ -1,10 +1,13 @@
 public class MopTrigger : PickUpTrigger
 {
-    public override bool Interact (GameObject originator, Vector3 position, string hand)
+	private MetaDataLayer metaDataLayer;
```
forgot to remove this field

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterPlayer.cs:
```diff
@@ -118,4 +136,71 @@ public void GetUp()
 	}
+
+	//Handled via UpdateManager
+	void UpdateMe()
```
I still want you to do this with a coroutine on server instead of introducing all these ticks/tickrates and writing extra code just for slipping. Players shouldn't have these tick loops running every frame if they're not stunned.

fomalsd commented on UnityProject/Assets/Scripts/Player/CrossedBehaviors/Slippery.cs:
```diff
@@ -0,0 +1,33 @@
+			return;
+		}
+		registerPlayer.Stun();
+		SoundManager.PlayNetworkedAtPos("Slip", registerItem.Position);
```
```suggestion
		SoundManager.PlayNetworkedAtPos("Slip", registerItem.WorldPosition);
```

fomalsd commented on UnityProject/Assets/Scripts/Player/PlayerSync.Server.cs:
```diff
@@ -596,4 +600,42 @@ private void CheckMovementServer()
+
+	public void CheckTileSlip()
+	{
+		var worldPosition = serverState.WorldPosition.CutToInt();
```
unused var

fomalsd commented on UnityProject/Assets/Scripts/Health/Living/LivingHealthBehaviour.cs:
```diff
@@ -677,6 +679,12 @@ public void AddHealth(int amount)
 		}
 	}
+
+	/// <summary>
```
dangling summary

fomalsd reviewed (changes requested):
Still didn't fix this one:
`Banana peels mostly work as intended, an exception is that they make you slip even if they're in the air and sometimes make thrower slip if he's running and throwing in the same direction. I don't think that should be the case. Check if peel is being thrown (in the air) at the moment (customNetTransform.IsBeingThrown on serverside) before invoking slip action.`
![Scr 2019-03-30 в 14 04 25 2019-03-30 14_42_26](https://user-images.githubusercontent.com/10403536/55275979-87801600-52fe-11e9-87f2-664e2b221e21.gif)


Also following criteria isn't finished, you always slip on wet floors even if you walk slowly:
`Players should slip on wet floors if they run on such tile (shouldn't slip if walking)`

Itsnotjames committed:
Various changes

In MetaDataLayer.cs removed Debug.Log() that was left in.
In MopTrigger.cs removed metaDataLayer field.
In Slippery.cs changed the audio to play at a WorldPosition instead of a localPosition.
In PlayerSync.Server.cs removed the "worldPosition" definition as it was unused.
In LivingHealthBehaviour.cs removed dangling summary.
In RegisterPlayer.cs changed stun timer from using the update manager to a coroutine.
In RegisterPlayer.cs made the Slip() method and changed Stun()'s calls to call Slip(). Slip() contains logic to prevent slipping while walking then calls Stun().
In Slippery.cs added a check to Slip() that the item isn't being thrown before slipping the player.

Aranclanos committed:
Merge pull request #4 from Itsnotjames/hamishthree

Various changes

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterPlayer.cs:
```diff
@@ -137,19 +122,14 @@ public void GetUp()
-		//Server Only:
-		if (CustomNetworkManager.Instance._isServer)
+		// Don't slip while walking
+		if (!UIManager.WalkRun.running)
```
that will only work when self-hosting, otherwise all clients will always slip if serverplayer has running mode enabled.
instead, check if player's current speed (`playerScript.playerSync.SpeedServer`) 
is greater than his walking speed (`playerScript.playerMove.WalkSpeed`)

but! remember that players should _always_ slip on banana peels regardless of movement speed

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterPlayer.cs:
```diff
@@ -161,35 +141,21 @@ public void Stun(float stunDuration = 4f, bool dropItem = true)
-			{
-				RemoveStun();
-			}
+			yield return new WaitForSeconds(stunTime);
```
Yeah, that's pretty good

fomalsd reviewed (commented):
Just one fix left!

Itsnotjames committed:
Changes walking check in Slip(), and adds summaries to Slip() and Stun().

Slip() now checks if the player is walking by checking if the player's speed is greater than walking speed. Before this Slip() used the UIManager to check if the player is walking.
Added summaries and parameter descriptions to Slip() and Stun().
Added an option to slip while walking in Slip().
Players now always slip on bananas.

Aranclanos committed:
Merge pull request #5 from Itsnotjames/hamishthree

Changes walking check in Slip(), and adds summaries to Slip() and Stu…

fomalsd reviewed (changes requested):
a few small things revealed in playtest. 
please also pull from latest develop to get proper player facing updates on your branch

Itsnotjames committed:
Merge remote-tracking branch 'upstream/develop' into hamishthree

Itsnotjames committed:
Adds PlayerUprightMessage

Written by foma.

Co-Authored-By: fomalsd <fomalsd@gmail.com>

Itsnotjames committed:
Adds PlayerUprightMessage.cs.meta

Co-Authored-By: fomalsd <fomalsd@gmail.com>

Itsnotjames committed:
Makes various LayDown, GetUp, Slip, and Stun changes

Moves the slip sound call from Slippery.Slip() to RegisterPlayer.Slip(). This makes every slip play the slip noise. All slip sounds now also play for hosts and clients.
Adds checks for crit, soft crit, and death in RegisterPlayer.Slip(), now players cannot slip while in these states.
Makes players stop pulling their current item when they slip.
Makes RegisterPlayer.Stun() and UpdateCanMove() use PlayerUprightMessage instead of LayDown() and GetUp(). This makes getting up and laying down shown for clients.

Aranclanos committed:
Merge pull request #6 from Itsnotjames/hamishthree

Updates branch from develop, and makes various changes
